### PR TITLE
Use default attribute value instead of raising exception if relationship is none on Relationship field

### PR DESCRIPTION
**Pull request by akira-dev**
I can't compute a related relationship url if relationship is none and related_kwargs include a dot.
Example:

```python
    item = Relationship(related_view='item_detail',
                        related_view_kwargs={'id': '<item.identifier>'},
                        schema='ItemSchema',
                        id_field='identifier',
                        type_='item')
```
In this case relationship is none so my_object.item is None and I receive an error:
AttributeError: 'item.identifier' is not a valid attribute of <Item object>

But 'item.identifier' is a valid attribut of my_object but the relationship is not mandatory so my_object.item could be None.

So I would like to use the default attribute of a field to set the value return by the serialization of the relationship field instead of the exception.

Plus if the default value is None I would like to get the same behavior for a relationship field as any other field so the relationship links are not included to the result.

Do you think it is the right way or I missed something ?

### CODE REVIEW

**Comment by sloria** [comment]
Thanks @akira-dev . I think you are describing the desired behavior. Could you please add tests for this? Those will help me understand the behavior better.

**Comment by akira-dev** [comment]
Ok I have added test

**Comment by akira-dev** [comment]
If you want to see I made a Flask extension for building REST API according to JSONAPI 1.0 specification: https://github.com/miLibris/flask-rest-jsonapi with your library so great thanks to you.

**Comment by akira-dev** [comment]
Do you know when you will release the next version of this library with this bug fix / feature ?

**Comment by sloria** [comment]
We do not have a scheduled release date, but I'll hopefully get a chance to take another look at this over the weekend. If it gets merged, I'll make sure to cut a release.

File: marshmallow_jsonapi/schema.py [hunk 1]
```diff
@@ -275,7 +275,7 @@ def format_item(self, item):
                 if 'meta' not in ret:
                     ret['meta'] = self.dict_class()
                 ret['meta'].update(value)
-            elif isinstance(self.fields[attribute], BaseRelationship):
+            elif isinstance(self.fields[attribute], BaseRelationship) and value:
```
Comment (sloria) [inline marshmallow_jsonapi/schema.py]: Is this change necessary? It seems possibly incorrect. If `value` is `None` then the field will be added to `attributes` rather than `relationships`.

**Review by sloria — CHANGES_REQUESTED** [review]
One concern noted above. I'd also like to see a test that uses `Relationship` within a `Schema`, so that the change to `format_item` is covered.

**Comment by akira-dev** [comment]
You're right, I haven't seen the side effect. I fixed it and add test with schema.

**Comment by sloria** [comment]
This change will also need to be implemented (and tested) for `marshmallow_jsonapi.fields.Relationship`

**Comment by akira-dev** [comment]
I have implemented this change for ```marshmallow_jsonapi.fields.Relationship``` and it is tested

File: marshmallow_jsonapi/fields.py [hunk 2]
```diff
@@ -120,14 +121,22 @@ def schema(self):
 
     def get_related_url(self, obj):
         if self.related_url:
-            kwargs = resolve_params(obj, self.related_url_kwargs)
-            return self.related_url.format(**kwargs)
+            kwargs = resolve_params(obj, self.related_url_kwargs, default=self.default)
+            for key, value in copy(kwargs).items():
```
Comment (sloria) [inline marshmallow_jsonapi/fields.py]: Nitpick: Rather than copy the dictionary, this can be written more expressively as a dictionary comprehension:

```python
            params = resolve_params(obj, self.related_url_kwargs, default=self.default)
            non_null_params = {
                key: value for key, value in params.items()
                if value is not None
            }
            if non_null_params:
                return self.related_url.format(**non_null_params)
```

Comment (akira-dev) [reply marshmallow_jsonapi/fields.py]: You're right changes are made

**Comment by akira-dev** [comment]
I made the changes

**Comment by sloria** [comment]
Thanks @akira-dev for making all those changes. I'll do one last review of this later today. In the meantime, feel free to squash your changes. If you don't know how to or don't have the time to do that, don't worry about it.

**Comment by akira-dev** [comment]
You want I rebase the branch and squash all commits in one ?

**Comment by sloria** [comment]
You don't necessarily need to squash them into one commit; squashing just a few of the minor style fixes is fine.

**Comment by akira-dev** [comment]
Ok it is done.

**Review by sloria — APPROVED** [review]

**Comment by sloria** [comment]
Thanks!
